File: Baekjoon/Class3/26.11724 연결 요소의 개수.js
const inputFile = process.platform === "linux" ? "/dev/stdin" : "./input.txt";
const input = require("fs").readFileSync(inputFile, "utf8").trim().split("\n");

const countComponents = (graph) => {
    let count = 0;
    const visited = new Array(graph.length).fill(false);
    for(let i = 1; i < graph.length; i++) {
        if(visited[i]) continue;
        let bfsHead = 0; // 컴포넌트 발견
        const bfs = [i];
        visited[i] = true;
        count++;
        while(bfsHead < bfs.length) {
            const start = bfs[bfsHead++];
            for(const end of graph[start]) {
                if(!visited[end]) {
                    bfs.push(end);
                    visited[end] = true;
                }
            }
        }
    }
    return count;
}

const [n, m] = input[0].split(" ").map(Number);
const graph = Array.from({ length: n + 1 }, () => []);

for(let i = 1; i <= m; i++) { // 무방향성 그래프
    const [u, v] = input[i].split(" ").map(Number);
    graph[u].push(v);
    graph[v].push(u);
}

console.log(countComponents(graph));

File: Baekjoon/Class3/26.11724 연결 요소의 개수.py
from collections import deque
import sys
input = sys.stdin.readline

def count_components(graph):
    count = 0
    bfs = deque()
    visited = [False] * len(graph)
    for i in range(1, len(graph)):
        if visited[i]:
            continue
        bfs.append(i) #컴포넌트 발견
        visited[i] = True
        count += 1
        while bfs:
            start = bfs.popleft()
            for end in graph[start]:
                if not visited[end]:
                    bfs.append(end)
                    visited[end] = True
    return count

def main():
    n, m = map(int, input().split())

    graph = [[] for _ in range(n + 1)] #무방향성 그래프
    for _ in range(m):
        u, v = map(int, input().split())
        graph[u].append(v)
        graph[v].append(u)

    print(count_components(graph))

main()